### ISSUE TITLE
Upgrade from 1.5.1 to 20170930 Failure

### ISSUE THREAD

**Issue by rough007** [log2timeline/plaso#1484]
**Plaso version:**

Trying to upgrade from version 1.5.1 to 20170930

**Operating system Plaso is running on:**

Ubuntu 16.04.3 LTS

**Installation method:**

Installed 1.5.1 and attempted to do an upgrade to 20170930 following guide on this Github Wiki here https://github.com/log2timeline/plaso/wiki/Ubuntu-Packaged-Release

**Details of commands run (in order) for the update from 1.5.1 to 20170930**
sudo add-apt-repository -y universe
sudo add-apt-repository -y ppa:gift/stable
sudo apt-get -y purge python-plaso plaso-tools
sudo rm -rf /usr/lib/python2.7/dist-packages/plaso
sudo apt-get -y -f install
sudo apt -y autoremove
sudo apt-get -y update;sudo apt-get -y dist-upgrade
sudo apt-get -y install python-plaso plaso-tools
sudo shutdown -r "now"

## To confirm it is installed:
~$ log2timeline.py --version
plaso - log2timeline version 20170930
~$ psort.py --version
plaso - psort version 20170930

## Attempted to use troubleshooting guide and here are the results
~$ wget https://raw.githubusercontent.com/log2timeline/plaso/master/utils/check_dependencies.py
--2017-10-27 17:26:48--  https://raw.githubusercontent.com/log2timeline/plaso/master/utils/check_dependencies.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 636 [text/plain]
Saving to: ‘check_dependencies.py’

check_dependencies.py    100%[==================================>]     636  --.-KB/s    in 0s

2017-10-27 17:26:49 (188 MB/s) - ‘check_dependencies.py’ saved [636/636]

~$ python check_dependencies.py
Traceback (most recent call last):
  File "check_dependencies.py", line 11, in <module>
    import utils.dependencies  # pylint: disable=wrong-import-position
ImportError: No module named utils.dependencies

**Description of problem:**

## When running the following command it fails:
~$ log2timeline.py -d Results/test.plaso Results/artifacts/C/Windows/System32/config/SECURITY
Checking availability and versions of dependencies.
[FAILURE]       artifacts version: 20160918 is too old, 20161022 or later required.

For more information see the [troubleshooting
guide][https://github.com/log2timeline/plaso/wiki/Troubleshooting].

**Debug output/tracebacks:**

Please run the tool with "-d" to generate debug output, and include anything relevant.
Same as above

**Source data:**

Please describe the source used.

SECURITY Registry Hive from Windows 7 workstation

**Comment by joachimmetz** [comment]
> [FAILURE] artifacts version: 20160918 is too old, 20161022 or later required.

Per https://github.com/log2timeline/plaso/blob/20170930/plaso/dependencies.py#L16 this is a legit warning.

version of artifacts in the stable track of the GIFT PPA is 20170909

https://launchpad.net/~gift/+archive/ubuntu/stable/+packages

Can you run:
```
apt-cache madison artifacts
```

to which versions are available

**Comment by rough007** [comment]
Sure thing.

** Here are the results of that command: **
~$ apt-cache madison artifacts
N: Unable to locate package artifacts

**Comment by joachimmetz** [comment]
The package name might be python-artifacts, however in the linked issue I'd noticed:

```
The following packages have unmet dependencies:
python-artifacts : Depends: artifacts-data but it is not installed
python-plaso : Depends: plaso-data but it is not installed
```

Was this resolved?

**Comment by rough007** [comment]
No. This does not work. The linked issue is someone else that started me down this path. This issue is for the simplified example and the exact commands I provided.  The new version of plaso still doesn't run and I did run the command to asked for and provided the feedback

**Comment by rough007** [comment]
So here is the thing. I started with an existing 1.5.1 and followed those exact commands which completely purged the old version and followed the instructions for the new version install based on the wiki. Now is there something that is missing from those commands?  If so then I would like to know what was missed so I can use thebew version

**Comment by rough007** [comment]
And I apologise for not being able test since doing this from my pgine.  I will try to add that plaso-data package manually and report back what I find in a bit

**Comment by joachimmetz** [comment]
So I've upgraded 1.5.1 to 20170930 a couple of times recently. Depending on the system configuration different things can go wrong. From your description I derive that it seems the artifacts package has not been updated. This package now is split in artifacts-data and python-artifacts. So that we also can provide a python3-artifacts going forwards. Same for plaso, but that there is not a python3-plaso yet.

`sudo rm -rf /usr/lib/python2.7/dist-packages/plaso` this should not be necessary if solely apt (dpkg) has been used to install plaso. If you mix package managers (pip, dpkg) you can get all kinds of weird side effects. 

What is the output of:
`sudo apt-get install python-plaso plaso-tools`

**Comment by rough007** [comment]
Okay, back at a keyboard.

Here's the background on the linked issue.  The linked issue was a user of CCF-VM who used the update.sh script built into the CCF-VM.  That script includes the following lines that break Plaso when run:
```
echo "Updating OS"
sudo apt-get -y update
sudo apt-get -y dist-upgrade
sudo apt-get -y autoremove
```

This broke Plaso when updating from 1.5.1 to 20170930.  I attempted many times to find an update path in order to provide that back out to everyone and have failed.  

I created this ticket with a vastly simplified example to remove as many variables as possible in order to get help in finding a solution to why the commands from the wiki aren't resulting in a fully working Plaso v20170930. Did I screw them up, or is there a step that is not documented, or something else entirely?

I will update the linked issue on my side once this issue is resolved.  If you want, I'd love to provide the script to Plaso community directly for review/use as it might benefit everyone trying to upgrade.

## On the example I provided here are the results
~$ sudo apt-get install plaso-data
Reading package lists... Done
Building dependency tree
Reading state information... Done
plaso-data is already the newest version (20170930-1ppa1~xenial).
0 upgraded, 0 newly installed, 0 to remove and 6 not upgraded.

**Comment by rough007** [comment]
Should it show up at all in pip if installed with apt package manager?

**Comment by rough007** [comment]
I got it working with your help! Thank you!

You were right in that it was something to do with pip.  It turns out there were two pip installations keeping the old version of `artifacts` on the system.  

After some digging it was the following pip packages that had to be removed (in this specific order) in order to free up all dependencies on the `artifacts `package and allows the new plaso install to work. 
```
sudo -H pip uninstall PyYAML
sudo -H pip uninstall artifacts
```

## The final upgrade script for CCF-VM 2.x is:
```
sudo add-apt-repository -y universe
sudo add-apt-repository -y ppa:gift/stable
sudo apt -y purge python-artifacts python3-artifacts plaso plaso-data plaso-tools python-plaso forensics-all
sudo rm -rf /usr/lib/python2.7/dist-packages/plaso
sudo apt -y -f install
sudo apt -y autoremove
sudo apt -y autoclean
sudo -H pip uninstall PyYAML
sudo -H pip uninstall artifacts

sudo apt -y update;sudo apt-get -y dist-upgrade
sudo apt -y install python-plaso plaso-tools
sudo shutdown -r "now"
```

## The final upgrade script for servers without those issues is:
```
sudo add-apt-repository -y universe
sudo add-apt-repository -y ppa:gift/stable
sudo apt -y purge python-artifacts python3-artifacts plaso plaso-data plaso-tools python-plaso forensics-all
sudo apt -y -f install
sudo apt -y autoremove
sudo apt -y autoclean

sudo apt -y update;sudo apt-get -y dist-upgrade
sudo apt -y install python-plaso plaso-tools
sudo shutdown -r "now"
```

**Comment by joachimmetz** [comment]
Good to hear you got it working.

> Should it show up at all in pip if installed with apt package manager?

We recommend to stick with apt (GIFT PPA) and not to use pip. You can get plaso working with pip but using it in parallel with apt can cause version conflicts. If you really want to use pip use it in combination with virtualenv.

> Did I screw them up, or is there a step that is not documented, or something else entirely?

The process is very straight forward when all variables are favorable. However when something is not-favorable, troubleshooting is very configuration dependent, which will be impossible for us to document.

I'll close this issue. Also I'm not sure what the `forensics-all` package is for but it is not needed for plaso
